crates/red_knot_python_semantic/resources/mdtest/import/conflicts.md

```diff
@@ -0,0 +1,63 @@
+# Conflicting attributes and submodules
+
+## Via import
+
+```py
+import a.b
+
+reveal_type(a.b)  # revealed: <module 'a.b'>
+```
+
+```py path=a/__init__.py
+b = 42
+```
+
+```py path=a/b.py
+```
+
+## Via from/import
+
+```py
+from a import b
+
+reveal_type(b)  # revealed: Literal[42]
+```
+
+```py path=a/__init__.py
+b = 42
+```
+
+```py path=a/b.py
+```
+
+## Via both
+
+```py
+import a.b
+from a import b
+
+reveal_type(b)  # revealed: <module 'a.b'>
+```
+
+```py path=a/__init__.py
+b = 42
+```
+
+```py path=a/b.py
+```
+
+## Via both (backwards)
+
+```py
+from a import b
+import a.b
+
+reveal_type(b)  # revealed: <module 'a.b'>
```

Comment:
> `b` is always `Literal[42]` in there

Not always; not if someone else in the process has done `import a.b` first. Though we intentionally don't model this global action-at-a-distance.

This test reflects the other known limitation of our model (currently), that our treatment of "is the submodule imported" is not flow-sensitive; if `a.b` is imported, we always treat it as if it were imported first. This is clearly documented elsewhere, but I guess to avoid confusion it could be mentioned in this test too (at least to point the reader to where it is discussed.)

So it is intentional here to test the type of `b`. (Though it wouldn't hurt also test `a.b` type.)

Reply:
Added a test of `a.b` as well, and added some commentary describing how this behavior differs from the interpreter